### PR TITLE
Fix add stroke folders as metadata

### PR DESCRIPTION
**Pull request by IntegralProgrammer**
Updates the _master_ branch with features necessary for the _StrokeLab_

### CODE REVIEW

File: image-archive/reactive-search/pages/index.jsx [hunk 1]
```diff
@@ -410,15 +411,15 @@ class Main extends Component {
...
-        lastquery = JSON.stringify({ query: lastquery, aggs: "" });
-        break;
-      }
-    }
+
+    //from _msearch, the query containing all the query terms, etc. comes after the index containing {preference: results}
+    var thisindex = lastquery.indexOf(`{"preference":"results"}`);
+
+    lastquery = JSON.parse(lastquery[thisindex + 1]).query;
+    lastquery = JSON.stringify({ query: lastquery, aggs: "" });
```
Comment (danielsnider) [inline image-archive/reactive-search/pages/index.jsx]: This is fine for now! I hope we eventually find a less hacky way though 🙂👍.

File: image-archive/de-id/deid_dicoms.py [hunk 2]
```diff
@@ -1506,7 +1511,10 @@ def insert_dicom_into_elastic(dicom):
   finished_indexing_doc = False
   while not finished_indexing_doc:
     try:
-      res = es.index(body=dicom_metadata, index=INDEX_NAME)
+      #res = es.index(body=dicom_metadata, index=INDEX_NAME)
+      log.info('Trying to insert into ElasticSearch: ' + str(dicom_metadata))
+      res = es.index(body=dicom_metadata, index=INDEX_NAME, doc_type='_doc') #Necessary for this to work on HPF.
+      log.info('Sucessfully inserted into ElasticSearch: ' + str(dicom_metadata))
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Fixed :heavy_check_mark:

File: image-archive/de-id/deid_dicoms.py [hunk 3]
```diff
@@ -1506,7 +1511,10 @@ def insert_dicom_into_elastic(dicom):
   finished_indexing_doc = False
   while not finished_indexing_doc:
     try:
-      res = es.index(body=dicom_metadata, index=INDEX_NAME)
+      #res = es.index(body=dicom_metadata, index=INDEX_NAME)
+      log.info('Trying to insert into ElasticSearch: ' + str(dicom_metadata))
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Fixed :heavy_check_mark:

File: image-archive/environments/strokelab/ingest_new_images_on_ready.py [hunk 4]
```diff
@@ -68,20 +75,21 @@ def __init__(self, dicom_paths):
     for dicom_path in dicom_paths:
       dicom = pydicom.dcmread(dicom_path, force=True)
       date = get_best_study_date(dicom)
-      if date is None:
+      #if date is None:
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/environments/strokelab/ingest_new_images_on_ready.py]: Fixed :heavy_check_mark:

File: image-archive/environments/strokelab/ingest_new_images_on_ready.py [hunk 5]
```diff
@@ -188,13 +200,21 @@ def main():
           else:
             shutil.copy(dicom_path, output_filepath)
           log.info('Output filepath: %s' % output_filepath)
+          cmd = ['qsub', '-F', output_filepath, '~/aim-platform/image-archive/environments/strokelab/qsub-deid-StrokeLab.sh']
+          log.info('Executing qsub ingestion: %s' % " ".join(cmd))
+          os.system(" ".join(cmd))
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/environments/strokelab/ingest_new_images_on_ready.py]: Not necessary, `os.system` was used to fix the problem associated with `Popen`.

File: image-archive/de-id/deid_dicoms.py [hunk 6]
```diff
@@ -386,6 +386,11 @@ def dicom_to_dict_for_elastic(dicom, log=None, environ=None):
   #   if not isinstance(v,str) and '__str__' in dir(v):
   #     dicom_metadata[k] = dicom_metadata[key].__str__()
 
+  #Force the copying of AccessionNumber, SeriesNumber, SeriesDescription
+  dicom_metadata['AccessionNumber'] = dicom.get('AccessionNumber')
+  dicom_metadata['SeriesNumber'] = dicom.get('SeriesNumber')
+  dicom_metadata['SeriesDescription'] = dicom.get('SeriesDescription')
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: This is needed due to the spacing issue in _pydicom_ raised by the issue (https://github.com/pydicom/pydicom/issues/977)

File: image-archive/de-id/deid_dicoms.py [hunk 7]
```diff
@@ -1506,7 +1511,10 @@ def insert_dicom_into_elastic(dicom):
   finished_indexing_doc = False
   while not finished_indexing_doc:
     try:
-      res = es.index(body=dicom_metadata, index=INDEX_NAME)
+      #res = es.index(body=dicom_metadata, index=INDEX_NAME)
+      log.info('Trying to insert into ElasticSearch: ' + str(dicom_metadata))
+      res = es.index(body=dicom_metadata, index=INDEX_NAME, doc_type='_doc') #Necessary for this to work on HPF.
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Fixed :heavy_check_mark:

File: image-archive/de-id/deid_dicoms.py [hunk 8]
```diff
@@ -1622,7 +1630,7 @@ def insert_deid_report_into_elastic(dicom):
   report['original_title'] = 'Report'
 
   # Insert
-  res = es.index(body=report, index=DEID_REPORT_INDEX_NAME)
+  res = es.index(body=report, index=DEID_REPORT_INDEX_NAME, doc_type='_doc') #Necessary for this to work on HPF.)
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Fixed :heavy_check_mark:

File: image-archive/de-id/deid_dicoms.py [hunk 9]
```diff
@@ -1653,7 +1661,8 @@ def generate_uid(dicom_dict, function_name=None, field_name=None):
       query = {'orig': orig, 'new': uid, 'field': field_name, 'date': datetime.datetime.now(), 'orig_path': dicom_dict['orig_path'], 'new_path': dicom_dict['new_path']}
       if '_id' in dicom_dict:
         query['id'] = dicom_dict['_id'] # Use same elasticsearch document ID from input index in the new index
-      res = es.index(body=query, index=LINKING_INDEX_NAME) #, doc_type=LINKING_DOC_TYPE)
+      #res = es.index(body=query, index=LINKING_INDEX_NAME) #, doc_type=LINKING_DOC_TYPE)
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Fixed :heavy_check_mark:

File: image-archive/de-id/deid_dicoms.py [hunk 10]
```diff
@@ -1653,7 +1661,8 @@ def generate_uid(dicom_dict, function_name=None, field_name=None):
...
+      res = es.index(body=query, index=LINKING_INDEX_NAME, doc_type='_doc') #Necessary for this to work on HPF.
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Fixed :heavy_check_mark:

File: image-archive/de-id/deid_dicoms.py [hunk 11]
```diff
@@ -2051,16 +2060,31 @@ def add_redcap_clinical_data(dicom):
...
+def parse_dicom_stroke_filepath(dicom_fp):
+  base_folder = dicom_fp.split('/')[0]
+  results = re.search(r'([^-_\/]+)_([A-Z\d]+)_([\d-]+)_(\d+)_(SE\d+)', dicom_fp)
+  study_name = results.group(1)
+  site_code = results.group(2)
+  subject_id = results.group(3)
+  visit_number = results.group(4)
+  session_number = results.group(5)
+  instance_number = os.path.basename(dicom_fp)
+  return base_folder, study_name, site_code, subject_id, visit_number, session_number, instance_number
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Thanks! :+1:

File: image-archive/de-id/deid_dicoms.py [hunk 12]
```diff
@@ -2051,16 +2060,31 @@ def add_redcap_clinical_data(dicom):
...
+  Example input: IPSS_Only/IPSS_049_312-50/IPSS_049_312-50_10_SE01_MR/009-t2-flair-ax-fs/100885078.dcm
+   - IPSS_Only (base_folder),
+   - IPSS (study_name),
+   - 049 (Site Code),
+   - 312-50 (subject_id),
+   - 10 (visit_number),
+   - SE01 (session_number),
+   - 009-t2-flair-ax-fs (study_description),
+   - 100885078.dcm (instance_number)
+  """
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Thanks :+1:

File: image-archive/de-id/deid_dicoms.py [hunk 13]
```diff
@@ -2284,9 +2309,9 @@ def store_number_of_redacted_PHI(dicom_uuid):
   """Record how many PHI was found and replaced in ElasticSearch"""
   if args.save_to_elastic:
     query = {'type': 'pixel_PHI', 'count': found_PHI_count_pixels, 'uuid': dicom_uuid}
-    res = es.index(body=query, index=COUNT_INDEX_NAME)
+    res = es.index(body=query, index=COUNT_INDEX_NAME, doc_type='_doc') #Necessary for this to work on HPF.
     query = {'type': 'header_PHI', 'count': found_PHI_count_header, 'uuid': dicom_uuid}
-    res = es.index(body=query, index=COUNT_INDEX_NAME)
+    res = es.index(body=query, index=COUNT_INDEX_NAME, doc_type='_doc') #Necessary for this to work on HPF.
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: Fixed :heavy_check_mark:

File: image-archive/environments/production/.tmuxinator.yml [hunk 14]
```diff
@@ -10,8 +10,8 @@ windows:
       layout: tiled
       root: <%= ENV["WORKSPACE"] %>/elastic-search
       panes:
-        - sudo docker logs -f elasticsearch
-        - sudo docker start elasticsearch
+        - docker logs -f elasticsearch
+        - docker start elasticsearch || docker run -d --name elasticsearch -p 9200:9200 -p 9300:9300 -e "ES_JAVA_OPTS=-Xms3g -Xmx3g" -e "discovery.type=single-node" -v `pwd`/elasticsearch.yml:/usr/share/elasticsearch/config/elasticsearch.yml -v $HOME/strokelab_esdata:/usr/share/elasticsearch/data docker.elastic.co/elasticsearch/elasticsearch:7.4.2
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/environments/production/.tmuxinator.yml]: Thanks! :+1:

File: image-archive/environments/strokelab/qsub-deid-StrokeLab-alternate.sh [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+cd ~/aim-platform/image-archive/de-id/ # important (some paths are not absolute :-())
+python ~/aim-platform/image-archive/de-id/deid_dicoms.py \
+  --input_filelist $INPUT_FILE_LIST.txt \
+  --input_base_path "/hpf/projects/ndlamini/images/inprogress/" \
+  --output_folder "/hpf/projects/ndlamini/images/archive/" \
+  --no_duplicate \
+  --no_deidentify \
+  --stroke_folders_as_metadata \
+  --redcap_csv_filepath "~/ipss_v3_all_data_20190918.csv" \
+  --deid_recipe "/home/dsnider/aim-platform/image-archive/de-id/deid.recipe"
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/environments/strokelab/qsub-deid-StrokeLab-alternate.sh]: This file, `qsub-deid-StrokeLab-alternate.sh`, was removed as it is no longer necessary. It is not called by any other program.

File: image-archive/environments/strokelab/qsub-deid-StrokeLab-multi.sh [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+  --input_filelist $INPUT_FILE \
+  --input_base_path "/hpf/projects/ndlamini/images/" \
+  --output_folder "/hpf/projects/ndlamini/images/archive/" \
+  --no_duplicate \
+  --no_deidentify \
+  --stroke_folders_as_metadata \
+  --redcap_csv_filepath "~/ipss_v3_all_data_20190918.csv" \
+  --deid_recipe "/home/dsnider/aim-platform/image-archive/de-id/deid.recipe"
+
+  # --input_base_path "/hpf/projects/ndlamini/images/inprogress/" \
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/environments/strokelab/qsub-deid-StrokeLab-multi.sh]: The only differences between `qsub-deid-StrokeLab-multi.sh` and `qsub-deid-StrokeLab.sh` is that `qsub-deid-StrokeLab-multi.sh` takes in a list of DICOM files and **does not** de-identify them prior to insertion into Image Archive. `qsub-deid-StrokeLab.sh`, which is called by `ingest_new_images_on_ready.py`, takes in a _single_ DICOM image, de-identifies it, and inserts it into Image Archive.

File: image-archive/reactive-search/pages/CollapsedViewSelector.jsx [hunk 17]
```diff
@@ -133,7 +137,26 @@ class CustomComponent extends Component {
...
+        'X-Requested-With': AUTH_TOKEN
+      }
     })
+    .then((response) => response.json())
+    .then((response) => {
+      this.setState({
+        totalCount: response.count.toLocaleString(),
+        totalCountFetched: true
+      });
+    });
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/reactive-search/pages/CollapsedViewSelector.jsx]: Thanks! :+1:

File: image-archive/reactive-search/pages/Leftbar.jsx [hunk 18]
```diff
@@ -343,6 +348,61 @@ const components = {
...
+    multiListRedCapRepeatInstance: {
+    react: {
+      and: searchBasedOn("redcap_repeat_instance-list"),
+    },
+    ...defaultMultiList,
+    componentId: "redcap_repeat_instance-list",
+    dataField: "redcap_repeat_instance.raw",
+    className: "redcap_repeat_instance-filter",
+    placeholder: "Search for a RedCap Repeat Instance",
+    filterLabel: "RedCap Repeat Instance"
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/reactive-search/pages/Leftbar.jsx]: Thanks! :+1:

File: image-archive/reactive-search/pages/Leftbar.jsx [hunk 19]
```diff
@@ -548,6 +613,36 @@ class Leftbar extends Component {
...
+          <LeftbarComponent icon="fa fa-repeat" title="RedCap Repeat Instrument">
+            <MultiList {...components.multiListRedCapRepeatInstrument} />
+          </LeftbarComponent>
+        ) : null}
+          
+        {this.state.RedcapEnabled ? (
+          <LeftbarComponent icon="fa fa-repeat" title="RedCap Repeat Instance">
+            <MultiList {...components.multiListRedCapRepeatInstance} />
+          </LeftbarComponent>
+        ) : null}
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/reactive-search/pages/Leftbar.jsx]: Thanks! :+1:

File: image-archive/reactive-search/pages/index.jsx [hunk 20]
```diff
@@ -353,7 +354,7 @@ const components = {
       and: searchBasedOn("results")
     },
     className: "Result_card",
-    size: 110,
+    size: 25,
```
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/reactive-search/pages/index.jsx]: Thanks! :+1:

File: image-archive/reactive-search/pages/index.jsx (hunk 1 above)
Comment (IntegralProgrammer) [reply image-archive/reactive-search/pages/index.jsx]: Okay :ok:

File: image-archive/de-id/deid_dicoms.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (IntegralProgrammer) [reply image-archive/de-id/deid_dicoms.py]: @danielsnider Please see the latest version this branch and the response to https://github.com/pydicom/pydicom/issues/977 Changing `field.name` to `field.keyword` and ensuring that the field name is not an empty string seems to fix the problem and explicit copying of `AccessionNumber`, `SeriesNumber` and `SeriesDescription` are no longer necessary.

**Pull request merged**